apps/hermes/client/js/src/HermesConnection.ts

```diff
@@ -0,0 +1,198 @@
+import { EncodingType, PriceFeedMetadata, PriceUpdate } from "./types";
+import EventSource from "eventsource";
+
+const DEFAULT_TIMEOUT: DurationInMs = 5000;
+const DEFAULT_HTTP_RETRIES = 3;
+
+export type UnixTimestamp = number;
+export type DurationInSeconds = number;
+export type HexString = string;
+export type DurationInMs = number;
+
+export type HermesConnectionConfig = {
+  /* Timeout of each request (for all of retries). Default: 5000ms */
+  timeout?: DurationInMs;
+  /**
+   * Number of times a HTTP request will be retried before the API returns a failure. Default: 3.
+   *
+   * The connection uses exponential back-off for the delay between retries. However,
+   * it will timeout regardless of the retries at the configured `timeout` time.
+   */
+  httpRetries?: number;
+};
+
+export class HermesConnection {
+  private baseURL: string;
+  private timeout: DurationInMs;
+  private httpRetries: number;
+
+  /**
+   * Constructs a new Connection.
+   *
+   * @param endpoint endpoint URL to the price service. Example: https://website/example/
+   * @param config Optional HermesConnectionConfig for custom configurations.
+   */
+  constructor(endpoint: string, config?: HermesConnectionConfig) {
+    this.baseURL = endpoint;
+    this.timeout = config?.timeout || DEFAULT_TIMEOUT;
+    this.httpRetries = config?.httpRetries || DEFAULT_HTTP_RETRIES;
+  }
+
+  private async httpRequest<ResponseData>(
+    url: string,
+    options?: RequestInit,
+    retries = this.httpRetries,
+    backoff = 300
+  ): Promise<ResponseData> {
+    const controller = new AbortController();
+    const { signal } = controller;
+    options = { ...options, signal }; // Merge any existing options with the signal
+
+    // Set a timeout to abort the request if it takes too long
+    const timeout = setTimeout(() => controller.abort(), this.timeout);
+
+    try {
+      const response = await fetch(url, options);
+      clearTimeout(timeout); // Clear the timeout if the request completes in time
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+      return await response.json();
+    } catch (error) {
+      clearTimeout(timeout);
+      if (
+        retries > 0 &&
+        !(error instanceof Error && error.name === "AbortError")
+      ) {
+        // Wait for a backoff period before retrying
+        await new Promise((resolve) => setTimeout(resolve, backoff));
```

Comment:
Don't do this, but a cool esoteric trick is that in Node you can actually do this:

```ts
import { promisify } from 'util';

await promisify(setTimeout)(backoff);
```

The reason this works is that Node will actually promisify [using the `nodejs.util.promisify.custom` Symbol attached to a function](https://github.com/nodejs/node/blob/0281e2cbf0531ade992265b7b734dd3f1ecffe8a/lib/internal/util.js#L401-L410) if it exists, and [`setTimeout` defines that symbol](https://github.com/nodejs/node/blob/0281e2cbf0531ade992265b7b734dd3f1ecffe8a/lib/timers.js#L172-L180) using [a promisifed version of `setTimeout`](https://github.com/nodejs/node/blob/0281e2cbf0531ade992265b7b734dd3f1ecffe8a/lib/timers/promises.js#L52-L93).

There are actually good reasons to prefer this (or just use `import { setTimeout } from 'timers/promises'` directly) -- the promisified version does some things like handle  abort controllers, does not wrap the underlying `setTimeout` and is thus arguably marginally more performant, etc.  However, I don't think that either the `promisify` trick or the `timers/promises` module are portable so I wouldn't use it in a place like this.  Still, it's a cool trick to be aware of.